Make utils_cv pip-installable

simonzhaoms opened the pull request:
### Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->


### Related Issues
<!--- If it fixes an open issue, please link to the issue here. -->


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [x] I have updated the documentation accordingly.

simonzhaoms commented:
Test results:

```console
$ conda create -n test-cvbp python=3
$ conda activate test-cvbp
$ pip install -e git://github.com/microsoft/ComputerVision.git@refs/pull/286/head#egg=utils_cv
(test-cvbp) $ python
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import utils_cv
>>> 
```

Some questions needs to be clarified:
* Supported Python version: Python 3?
* Version scheme: `2019.08`?
* Package name: `utils_cv`?
* Author name: `CVDev Team at Microsoft`?

miguelgfierro commented on setup.py:
```diff
@@ -0,0 +1,60 @@
+    install_requires=[
+        "azureml-sdk[notebooks,contrib]>=1.0.30",  # requires ipykernel, papermill, jupyter-core, jupyter-client
+        "bqplot",
+        "fastai==1.0.48",  # requires pytorch, torchvision, nvidia-ml-py3
```
the main difference I see with what we are doing in recommenders or nlp is that there we are installing different versions for cpu and gpu, so our installation process is more complicated. Your installation dependencies are much simpler. I wonder what happens if you install this library in a new environment on a GPU machine, does it automatically install pytorch gpu?

simonzhaoms replied:
Yeah, currently it is much simpler for CVBP than Reco.  I am not sure if we need to separate it into CPU version, GPU version and Spark version, such as `utils_reco_cpu`, `utils_reco_gpu` and `utils_reco_spark`, and do similar thing for `utils_cv`.  That needs to be fully tested.  The only thing I know for PyTorch said in [fast.ai installation](https://github.com/fastai/fastai/blob/master/README.md#installation) is 
> Starting with pytorch-1.x you no longer need to install a special pytorch-cpu version. Instead use the normal pytorch and it works with and without GPU. But you can install the cpu build too.

miguelgfierro replied:
ok great

PatrickBue replied:
Yes. Our code and installation instructions is identical for CPU and GPU targets.

miguelgfierro commented on README.md:
````diff
@@ -116,3 +116,22 @@ To opt out of tracking, please go to the raw `.ipynb` files and remove the follo
+It is also possible to install directly from GitHub.
+
+```
+pip install git+https://github.com/microsoft/ComputerVision#egg=utils_cv
````
is this enough? or you need `pip install git+http://github.com/microsoft/ComputerVision/#egg=pkg&subdirectory=utils_cv`?

simonzhaoms replied:
Yes.

miguelgfierro commented on README.md:
````diff
@@ -116,3 +116,22 @@ To opt out of tracking, please go to the raw `.ipynb` files and remove the follo
+installation of this utilities in this repo from the main directory.
+
+```
+pip install -e .
````
if you do this, I'm not sure if it will add the tests, tools and notebooks to the library, I would try `pip install -e utils_cv`

simonzhaoms replied:
It would only add `utils_cv`.

miguelgfierro replied:
perfect

PatrickBue commented on utils_cv/README.md:
```diff
@@ -0,0 +1,24 @@
+# Recommender Utilities # 
```
[minor] old reference to Recommender

simonzhaoms replied:
:joy:

PatrickBue commented on setup.py:
```diff
@@ -0,0 +1,60 @@
+    ],
+    include_package_data=True,
+    # Not fully test the dependency coverage, please make sure to sync
+    # with environment.yml
```
Rather than explicitely listing all dependencies here, is it possible to refer to the environment.yml file? That way we also make sure that these two places stay in sync (also because the environment.yml file is used in our tests)

jiata replied:
instead of using our enviornment.yml, it _might_ be best to have the environment.yml use the utils_cv package, as so:

```yml
# environment.yml
name: cv
channels:
- defaults
...
dependencies:
- python==3.6.8
- pytorch==1.0.0
- torchvision>=0.3.0
- fastai==1.0.48
- ..
- pip:
  - azureml-sdk[notebooks,contrib]>=1.0.30
  - utils_cv>=0.0.1 <-- [install our lib here]
  - ...
```

simonzhaoms replied:
I agree to add `utils_cv` in environment.yml.  That would make sense.

simonzhaoms replied:
@PatrickBue @jiata See 63daec9db29f3f64b9e804afb4ac091b2552ff91
https://github.com/microsoft/ComputerVision/blob/63daec9db29f3f64b9e804afb4ac091b2552ff91/environment.yml#L12-L28

PatrickBue replied:
Please update/revert this setup script to reflect that the environment.yml file will stay unchanged.

simonzhaoms replied:
Reverted.

jiata commented on setup.py:
```diff
@@ -0,0 +1,60 @@
+    # Not fully test the dependency coverage, please make sure to sync
+    # with environment.yml
+    install_requires=[
+        "azureml-sdk[notebooks,contrib]>=1.0.30",  # requires ipykernel, papermill, jupyter-core, jupyter-client
```
do we actually want to put the azureml sdk into our lib?

simonzhaoms replied:
`utils_cv/common/deployment.py` and `utils_cv/common/azureml.py` depends on `azureml-sdk`.

jiata replied:
oh right. arguably we should not support these. but maybe we can deal with that in the future... 

But i think it doesn't make sense that we're providing these additions on top of azureml's sdk. If we are modifying the user experience of azureml because it is not good enough as is, that code should live in the azureml sdk... otherwise, we should stick to the way azureml docs recommends.

simonzhaoms replied:
Yeah, we can deal with that in the future.

jiata commented:
> Test results:
> 
> ```
> $ conda create -n test-cvbp python=3
> $ conda activate test-cvbp
> $ pip install -e git://github.com/microsoft/ComputerVision.git@refs/pull/286/head#egg=utils_cv
> (test-cvbp) $ python
> Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
> [GCC 7.3.0] :: Anaconda, Inc. on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import utils_cv
> >>> 
> ```
> 
> Some questions needs to be clarified:
> 
> * Supported Python version: Python 3?
> * Version scheme: `2019.08`?
> * Package name: `utils_cv`?
> * Author name: `CVDev Team at Microsoft`?

we probably wanna specify python version >= 3.6

simonzhaoms commented on setup.py:
```diff
@@ -0,0 +1,60 @@
+        "Topic :: Software Development :: Libraries :: Python Modules",
+    ],
+    include_package_data=True,
+    # Not fully test the dependency coverage, please make sure to sync
```
`azureml-sdk` depends on `papermill`.  `nteract-scapbook` is not used in `utils_cv`.

miguelgfierro replied:
coold

simonzhaoms commented:
> ### Description
> ### Related Issues
> ### Checklist:
> 
>     * [ ]  I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
> 
>     * [ ]  I have added tests covering my contributions.
> 
>     * [x]  I have updated the documentation accordingly.

BTW, The `contribution guidelines` in the template is broken.  And it may be : https://github.com/microsoft/ComputerVision/blob/master/CONTRIBUTING.md ?

PatrickBue commented on setup.py:
```diff
@@ -0,0 +1,62 @@
+        "azureml-sdk[notebooks,contrib]>=1.0.30",  # requires ipykernel, papermill, jupyter-core, jupyter-client
+        "bqplot",
+        "fastai==1.0.48",  # requires pytorch, torchvision, nvidia-ml-py3
+        "scikit-learn>=0.19.1",
```
Q1: How come there are only 4 libraries getting installed here, while in our environment.yaml file there are 8 pip-installable libraries?
Q2: so the user installs via "pip install git+https://github.com/microsoft/ComputerVision#egg=utils_cv" and then has to execute the "conda env create -f environment.yml" to create all dependencies?
Q3: If yes, can we just remove the 4 pip packages here. Also, if yes, this needs to be made clear in the text.

simonzhaoms replied:
@PatrickBue 

### A1 ### 
* Some libraries are required by these 4
  - E.g. `fast.ai` will install `pytorch`, `torchvision` and `nvidia-ml-py3`
* Some others are not depended by `utils_cv`
  - E.g. `ipywebrtc` which is used in notebook, and `black` is for reformatting.
https://github.com/microsoft/ComputerVisionBestPractices/blob/f194c2789349bf0be4d9369fe0d5cdba863bb420/environment.yml#L18-L37

### A2 ###
We can put `pip install git+https://github.com/microsoft/ComputerVision#egg=utils_cv` inside `environment.yml`.  E.g. 
<pre>
# environment.yml
name: utils_cv
channels:
- defaults
...
- dependencies:
  - pytest>=3.6.4
- pip:
  - <b>git+https://github.com/microsoft/ComputerVisionBestPractices#egg=utils_cv</b>
  - black>=18.6b4
  - ipywebrtc
  - pre-commit>=1.14.4
</pre>

### A3 ###
Yes, sure.

simonzhaoms replied:
@PatrickBue See 63daec9db29f3f64b9e804afb4ac091b2552ff91
https://github.com/microsoft/ComputerVision/blob/63daec9db29f3f64b9e804afb4ac091b2552ff91/environment.yml#L12-L28

PatrickBue replied:
A1 sounds good. A2 see my other comments. Please undo all edits to the environment.yaml file. A3 after reverting the environment.yaml file back to its unedited version, please re-add all libraries as required.

simonzhaoms replied:
Done.

PatrickBue commented on README.md:
````diff
@@ -116,3 +116,22 @@ To opt out of tracking, please go to the raw `.ipynb` files and remove the follo
 ```
 
+
+## Install this repository via PIP
````
Just adding this section to the end seems the wrong place. Maybe consider moving to the end of the "getting started" instructions, and condense all the text here into 3-4 sentences including when one would might want to do "pip install git+https://github.com/microsoft/ComputerVision#egg=utils_cv"

simonzhaoms replied:
Done in the new commit.

simonzhaoms commented:
> > Test results:
> > ```
> > $ conda create -n test-cvbp python=3
> > $ conda activate test-cvbp
> > $ pip install -e git://github.com/microsoft/ComputerVision.git@refs/pull/286/head#egg=utils_cv
> > (test-cvbp) $ python
> > Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
> > [GCC 7.3.0] :: Anaconda, Inc. on linux
> > Type "help", "copyright", "credits" or "license" for more information.
> > >>> import utils_cv
> > >>> 
> > ```
> > 
> > 
> > Some questions needs to be clarified:
> > 
> > * Supported Python version: Python 3?
> > * Version scheme: `2019.08`?
> > * Package name: `utils_cv`?
> > * Author name: `CVDev Team at Microsoft`?
> 
> we probably wanna specify python version >= 3.6

@jiata In `environment.yml`, it's `==3.6.8`:
https://github.com/microsoft/ComputerVision/blob/f4174e472eb08c6ff6af89c4accb1acb784787de/environment.yml#L18-L19

jiata commented on README.md:
```diff
@@ -38,7 +38,17 @@ To get started:
-    
+
+
+The utilities in [`utils_cv`](utils_cv) is `pip`-installable from the
```
i think that this could be clearer -- that the environment.yml will install the utils_cv package, but if you would like to install it in a separate environment outside of cvbp, you can just use the `pip install` method

PatrickBue replied:
I am not convinced the way this is done here is the way to go. As I mentioned multiple times before, I think we should wait and then copy what the Recommenders folks do. I will leave comments on this PR today. For now, my current thinking is to (i) add the option to do a pip install of the utils_cv folder; (ii) but not change any of the existing files (such as the environment.yaml) and leave our current installation instructions as-is

PatrickBue commented on environment.yml:
```diff
@@ -13,25 +13,16 @@ name: cv
 - pip>=19.0.3
 - pip:
-  - azureml-sdk[notebooks,contrib]>=1.0.30
+  # NOTE: Make sure of syncing dependencies in `setup.py`
```
As I outlined in my other comments, please undo all changes in this file.

PatrickBue commented on setup.py:
```diff
@@ -0,0 +1,63 @@
+        ]
+    ),
+    packages=[UTILS_CV],
+    python_requires="==3.6.8",
```
In my opinion, 3.6 is sufficient, no need to specify 3.6.8

simonzhaoms replied:
Done.

PatrickBue commented on README.md:
```diff
@@ -38,7 +38,17 @@ To get started:
-    
+
+
+The utilities in [`utils_cv`](utils_cv) is `pip`-installable from the
```
Bit unclear. Maybe use this text here instead: "As an alternative to the steps above, and if one wants to instal only the 'utils_cv' library (without creating a new conda environment), this can be done by running 
```bash 
pip install git+https://github.com/microsoft/ComputerVision.git@master#egg=utils_cv 
```
or by downloading the repo and then running `pip install .` in the root directory.

simonzhaoms replied:
Changed.
